### PR TITLE
Improve support for reference-types

### PR DESCRIPTION
**Pull request by eqrion**
This is the Cranelift part of the work to add reference-types to Spidermonkey+Cranelift (see [bug 1574865](https://bugzilla.mozilla.org/show_bug.cgi?id=1574865)). I will post the Spidermonkey patches there soon.

I also have two additional commits that are blocked on bytecodealliance/wasmparser#168, but this is the majority of the work.

### CODE REVIEW

**Comment by eqrion** [comment]
I've found an issue through more extensive testing, this isn't ready for review.

**Comment by eqrion** [comment]
> I've found an issue through more extensive testing, this isn't ready for review.

That was a false alarm, I was getting a timeout in a non-standard configuration. Upon further investigation, it looks like it's just caused by a very slow implementation of barriers. So this should be ready for review.

File: cranelift-wasm/src/code_translator.rs [hunk 1]
```diff
@@ -1077,56 +1077,73 @@ pub fn translate_operator<FE: FuncEnvironment + ?Sized>(
...
+            environ.translate_table_fill(builder.cursor(), *table, dest, val, len)?;
+        }
+        Operator::TableInit {
+            segment,
+            table: table_index,
+        } => {
             // The WebAssembly MVP only supports one table and we assume it here.
-            let table_index = 0;
-            let table = state.get_table(builder.func, table_index, environ)?;
+            let table = state.get_table(builder.func, *table_index, environ)?;
```
Comment (fitzgen) [inline cranelift-wasm/src/code_translator.rs]: The comment above this line seems like it is now out of date.

**Comment by eqrion** [comment]
> Code looks good to me, but I think it should be possible to have some tests, right? I don't think we actually need to manufacture any reference types to have tests (which would require askign how we're going to manage their lifetimes) and can instead have test functions that take `r32` and `r64` parameters and put them in tables and all that. I think we should be able to do that even in the dummy environment.

That seems plausible to me. I'm not very familiar with the testing architecture, so I'll need to do some investigation. I'll try to do that today or tomorrow.

**Comment by eqrion** [comment]
I've now added a basic test for r32 and r64, removed the out-of-date comments, and updated the documentation on `StackMap` to more thoroughly explain the representation.

**Pull request merged**